pilot/pkg/serviceregistry/kube/controller/serviceexportcache.go

```diff
@@ -191,8 +199,39 @@ func (ec *serviceExportCacheImpl) ExportedServices() []exportedService {
 	return out
 }
 
+func (ec *serviceExportCacheImpl) Run(stop <-chan struct{}) {
+	select {
+	case <-ec.serviceExportCh:
+	case <-stop:
+		return
+	}
+	dInformer := ec.client.DynamicInformer().ForResource(mcs.ServiceExportGVR)
+	_ = dInformer.Informer().SetTransform(kube.StripUnusedFields)
+	if ec.opts.DiscoveryNamespacesFilter != nil {
+		ec.filteredInformer = informer.NewFilteredSharedIndexInformer(ec.opts.DiscoveryNamespacesFilter.Filter, dInformer.Informer())
+	} else {
+		ec.filteredInformer = informer.NewFilteredSharedIndexInformer(nil, dInformer.Informer())
+	}
+	// Register callbacks for events.
+	ec.registerHandlers(ec.filteredInformer, "ServiceExports", ec.onServiceExportEvent, nil)
+	go ec.filteredInformer.Run(stop)
+	kube.WaitForCacheSync(stop, ec.filteredInformer.HasSynced)
+	ec.started.Store(true)
+}
+
 func (ec *serviceExportCacheImpl) HasSynced() bool {
-	return ec.filteredInformer.HasSynced()
+	return true
+}
+
+func (ec *serviceExportCacheImpl) OnCRDEvent(name string) {
+	if name == mcs.ServiceExportGVR.Resource+"."+mcs.ServiceExportGVR.Group {
+		select {
+		case <-ec.serviceExportCh: // channel already closed
```

Comment:
I think this is racy. OnCRDEvent can be called multiple times in parallel. if they both enter L231 at the same time they will both try to close. This approach to "close if not already closed" only works with locking/single thread

Reply:
While rare the impact is a panic in istiod so its not great

Reply:
This is run in serial, please refer pilot/pkg/serviceregistry/kube/controller/controller.go#L412